From acbfc6033edc75c721b1be8259fff85aba560797 Mon Sep 17 00:00:00 2001
From: Narendhar Reddy <nare@qti.qualcomm.com>
Date: Wed, 16 Jul 2025 20:43:57 +0530
Subject: [PATCH] QSDK:DP:ath12k: fix memory leak in ppeds.

Currently in rmmod/insmod usecase, memory allocated
for tx_status and ppeds_rx_idx is not getting freed, leading to
out of memory issue during continuous rmmod/insmod.

Fix the leak by properly freeing the memory.

Fixes : I56018d1a5b (QSDK: DP: wifi: Package upgrade changes for datapath accelerator features)
patch dependency : 0337-01-QSDK-DP-wifi-Add-support-for-PPE-DS-Direct-switch-fe.patch
Signed-off-by: Narendhar Reddy <nare@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/ppe.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/ppe.c b/drivers/net/wireless/ath/ath12k/ppe.c
index 24bd4a1..ee84987 100644
--- a/drivers/net/wireless/ath/ath12k/ppe.c
+++ b/drivers/net/wireless/ath/ath12k/ppe.c
@@ -241,7 +241,7 @@ static void ath12k_ppeds_release_rx_desc(ppe_ds_wlan_handle_t *ppeds_handle,
 
 	if (unlikely(count > ab->dp->ppe.ppeds_rx_num_elem)) {
 		new_size = sizeof(u16) * count;
-		for (device_id = 0; device_id < ATH12K_MAX_SOCS; device_id++) {
+		for (device_id = 0; device_id < ag->num_hw; device_id++) {
 			tmp = krealloc(ab->dp->ppe.ppeds_rx_idx[device_id], new_size, GFP_ATOMIC);
 			if (!tmp) {
 				ath12k_err(ab, "ppeds: rx desc realloc failed for size %u\n",
@@ -274,7 +274,7 @@ static void ath12k_ppeds_release_rx_desc(ppe_ds_wlan_handle_t *ppeds_handle,
 		rx_bufs_reaped[device_id]++;
 	}
 
-	for (device_id = 0; device_id < ATH12K_MAX_SOCS; device_id++) {
+	for (device_id = 0; device_id < ag->num_hw; device_id++) {
 		if (!rx_bufs_reaped[device_id])
 			continue;
 
@@ -1283,7 +1283,7 @@ int ath12k_ppeds_attach(struct ath12k_base *ab)
 	*abptr = ab;
 	ath12k_info(ab, "PPEDS attach success\n");
 
-	for (i = 0; i < ATH12K_MAX_SOCS; i++) {
+	for (i = 0; i < ab->ag->num_hw; i++) {
 		ab->dp->ppe.ppeds_rx_idx[i] = kzalloc((sizeof(u16) * PPE_DS_TXCMPL_DEF_BUDGET),
 						      GFP_ATOMIC);
 		if (!ab->dp->ppe.ppeds_rx_idx[i]) {
@@ -1716,6 +1716,7 @@ void ath12k_dp_srng_ppeds_cleanup(struct ath12k_base *ab)
 
 	ath12k_dp_srng_cleanup(ab, &dp->ppe.ppe2tcl_ring);
 	ath12k_dp_srng_cleanup(ab, &dp->ppe.reo2ppe_ring);
+	kfree(dp->ppe.ppeds_comp_ring.tx_status);
 	ath12k_dp_srng_cleanup(ab, &dp->ppe.ppeds_comp_ring.ppe_wbm2sw_ring);
 }
 
-- 
2.34.1

